lessons/python/intro-nlp/intro_NLP.ipynb

```diff
@@ -0,0 +1,3910 @@
\nzB5z952KyOR8Q3cA76Wjb+gmd++UyK+E/pPChrsM57UiC++iG5p3xfwjAwn5nHYuV8+0lNJ3dB45\nv8tkQtqWmzxhUVnuxsDMvhkdc74VyezaU8hH0EuwjrH0M9w97kDGo1xB0fzmWILz1Qk/4KTVi62E\nC1XRO4ES7iIPITiM47w/f5+ZbUmIqNk70vFuYKq7v5DQfapMoYUMG5AUH/8Zwh+2HzHjQl5eIgsx\n778gLPoy4HwzO83dr46JpfXN5NqcSnBkriDM0+8O/Le735Igm3bhW5qcSEm+IY/0+HUXx6VhlXXM\nhbU3sCpJ0PIW4EW+gmWEkMmHY2+lDVDAQn6rXEqG3O/pB56wjiElqfsmjBj3tRB4cTvht/AxQvRQ\nPmvN7DhCYEJubUenOgOlTLVVDO/heFVtpW/Az0gRSx/JnkSId/8D7SseT0yQuwoYV6TfgcAoQvKt\nkdHzUYS77rkxuS8Q5ttXEpxquW0RYaVxfru3Ei7GTdH2aeDWAjqkzRQ6F9rXMBSRfTSl3CPAZrHX\nm1I4C2VT2jajx0MJq2Z3oUB2T8IF7aDosxxPWL37/QS5JwgpuudHso9SOJb9uwS/AISFXX8Hdu/m\n73PX6LN6JtoeAqYUkL2CkJ787GibR3DcPwB8M0+2hRCNM5UCayhiv6fvANtE27eB27p5Trun7Du3\nNuDLOf2JrRPIk92BEDl0XPR6G8JNQL7cNYTcQhOj7QxCBoCyz6fo+VazcW0V+pLCn7sh9rqxiz/6\nfGCT2OtNCPlP8uXuIMzRTo8uSP8grJ6My0ylPeHawuj5ouhP/6WY3HCCcfhLdMHKbaMK6FhKGupr\ngKcI4Ybn5bYEuaKGLSZ7MWE0VEwufzFWQ8K+K3OydDSCc5K+o9w+4FzgI9HzQheOVAvf8j7zjVuh\n31L0uA/hbvbDdHNBGCGa6XiCkTknunB9t4DsdGKGnRBxdTMwCHgiT7aRMILZOrd19Tnl7euUyjnF\neeQM5KikrcAxDxFWnt8H7Jj0uylDj9T/j0ptmhrqPRSNpY94gTDcz7ECeD5B7nvFOnT3c4Fzi0UY\neRiyLjOzcwkV1zamBbDYQqcYS8zskwTDAWG+vtAw/l46J0nbuEjM2iuzDQWeMLP76Tjv38mXQLgI\nnhA5wbvKKnqTmU2P6fkxOpe/nBo9Hk46ZltIn7AN8C0LqRMKhUamWvjmkf/HQjGYgfnv55GbNvsg\noYrY9Wb2vZS6F+J6wlqLByMdu2Jrwuglx1qC0VplYTEgEKLaCAblZdoXMDrJawPuMLOP0x4qezSh\nQl2pXEH4HmeTEBxBchz/VwnhzH9398cjZ3gH53UZgRypp9oqhZzFvYDoR/4z2ueq9yNEDf01QfaP\nwM6EP6cT7vjuJwzH8eREdWl0KJqfxsweIkwzePS6gXBnlh9NszVhXvo9kY73AF9x9+cS+n0QOMHb\nq5kdB3zVoxoOZrY/4TM5E/hm/FDgTE+o9VCCQ/1MQlGc3PqJmcCe3j3ndwNhKmWhuy+N5re38IQQ\nTuucN2o4IW/UfXlyHyJMs2xOWCA2njB1t2NCmzcQLtYHE6ZeVhGc4Lt045yKOqtjst8hhGPm8jUd\nQTD0ZxNCNI+P5J4C9vAU8/zWnjcrZ+QaaU8c59519bWqU0Ygxy4E31buhu8Nwn+gW2G+XeooQ1D/\nRI7QBYQfxHN0kb7WUiaqsxLSPERtvpdgCP5JcADf7Xm1lgtEj2xMMRzbdxnhYv5G9HoU8AtPCI+L\n7rCuJkw97ENwtB3ueY6zAuGbnfouhTRtWomZQqNjRgLb0dGoFgyHtbAC2L1whM0jhFW8t3lI8XwA\nYR765ATZZkLZw0c9lJQcB+zsCc7qtFgobn++pyvgg4X6Ghtz7bt7fiGkXEjoIZ5iTUwkP4rOn+md\naY5NaOt6QlGY6909vyRskp5Jd/lJ61dSRdXFHOq5NCFvkuxQrxgyBL0AMzuQ9gLuEwmRMTOjqZuu\njmsgzMcWrUplUZoHd/+fhPcepT0/zS5RSOUl3jmr5bWEUUtuZfEpwAHufmSeXKeQua7C6KKwvOsI\nU1xHuvuq2HtfiPqZSMdVt0MJiddKTiNejTZjbZ9EmE7akvA97kmIUU+6cMRz6EC4GHTKoWMhtXRr\nZBB28xDKeL/3QOhq1H8pBXySFieu8PaqZbmL4I6E1dc30nGqL6keQdJneo+XuTYiGmV+jDB9dj8h\nLPgGTyicFBm1HAMJC/vWufs3E2RT3axYSLnRShgpWaTHA4RIt6s8b+FiRaimA0Jb5TbCcHdPwnzk\nsxTIxEiY5xxGGCrPI5RLPC1lH/cV2J+rnTs7attIyCpKmL/+K2F64uVIl80S5B4hVqeW4IzLd8Lm\nO19foj0qZk5MriRHdcrPoeJt5p3XQCLnH+HP/bcCsnOAfWOv9yHZAX0b4e7x/EjncwkXwp76bZbi\nrH6GMIWzhOAXWk/waz1ImKo6o4utkAM69Wda4nk1EkKirwSWl3DcnXmvS42qS+1Qr9QmZ3EvwEI9\n1MFEKQ4Ii10SV+ESomGWm9nxhGmc0wkX8J/ntZmU5qHQ8DBVfppIp4/n70/gbELs+9VRn8cCP86T\nSeV89chRTXA4V4RqtBljtbuvNjPMbICHPEnbF5BNm0Pnw4R8RF8jTKENp3C2zIrjXVRsS+BmgmN1\nOoCZvY8wVXUl8Btv9/1sTM+RI1rXkUQpn2kqzGwQwX/xMUIoaaHMp/ERTgPBmI3NE7uCEAb9UyCe\nU2uFJ2cUTeVQryQyBL2DOYQf2E6EC9RSC6UmkyIJ+lmognUk8Gt3X2tmSRf4+LROLs3Dh5M6d/dT\noqcXWkhnkJifxsIKzM8ShvXxudoOc//u/kczayPMaxtwlLs/kSdTysWlN/FCZFSvA241szfonEk2\nN31wv4Xi9fEcOjPyG/SO1dMqUoylirS6++
```

Comment:
Use parentheses instead of brackets (except for the last one) to save a few extra list copies:
```python
# lower case
filtered_words = (token.lower() for token in tokens)
# remove stop words
filtered_words = (w for w in filtered_words if not w in stops)
# filter for short punctuation
filtered_words = [w for w in filtered_words if len(w) > 2]
```